File: python/x402/mechanisms/evm/settle_receipt.py
"""Shared settle receipt wait helpers for EVM facilitators."""

from __future__ import annotations

from collections.abc import Callable
from typing import Protocol

from ...pending_settlement_store import PendingSettlementStore
from ...schemas import SettleResponse
from .constants import ERR_SETTLEMENT_PENDING, TX_STATUS_SUCCESS
from .types import TransactionReceipt
from .utils import is_valid_tx_hash, truncate_error_message


class ReceiptWaiter(Protocol):
    """Signer capability required to confirm a broadcast settlement transaction."""

    def wait_for_transaction_receipt(self, tx_hash: str) -> TransactionReceipt:
        """Block until the transaction is mined and return its receipt."""
        ...


def invalid_broadcast_hash_response(
    tx: str,
    error_reason: str,
    network: str,
    payer: str | None = None,
) -> SettleResponse:
    """Terminal failure when a signer reports success without a usable hash."""
    return SettleResponse(
        success=False,
        error_reason=error_reason,
        error_message=f"signer returned an invalid transaction hash: {tx!r}",
        transaction="",
        network=network,
        payer=payer,
    )


def wait_for_receipt_and_build_response(
    signer: ReceiptWaiter,
    tx_hash: str,
    network: str,
    payer: str | None,
    *,
    failed_reason: str,
    amount: str | None = None,
    validate_receipt: Callable[[TransactionReceipt], SettleResponse | None] | None = None,
    on_success: Callable[[TransactionReceipt], SettleResponse] | None = None,
    pending_store: PendingSettlementStore | None = None,
    pending_key: str | None = None,
) -> SettleResponse:
    """Wait for a broadcast receipt and build the settlement response.

    validate_receipt runs after a successful receipt (e.g. a Transfer event check): return a
    SettleResponse to fail settlement, or None to accept success. on_success, when set,
    builds the success response from the receipt.

    When pending_store and pending_key are both provided, they key a store that a subsequent
    settle attempt for the same payload (typically the resource server's single automatic
    retry) uses to reconcile against tx_hash instead of re-broadcasting. Only a
    settlement_pending outcome (wait failure/timeout, or an exception while processing a
    confirmed receipt) is recorded — it is the only outcome safe to retry against. Every
    other outcome — success, an invalid hash, a reverted receipt, or an explicit
    validate_receipt rejection — is terminal and clears the entry instead; a reverted
    receipt still has a transaction hash but is not safe to reconcile against indefinitely,
    so it must not be cached until TTL expiry.

    If persisting the pending entry itself fails, a later retry has no record to reconcile
    against — blindly returning settlement_pending would let it re-verify/re-broadcast and
    risk a double-send. That case is downgraded to failed_reason, preserving tx_hash for
    manual reconciliation. A failure to clear the entry is swallowed instead: the settle
    outcome is already correct and must not be masked by a storage hiccup, and a stale entry
    merely lingers until TTL expiry.
    """

    def _mark_pending() -> SettleResponse | None:
        if pending_store is None or not pending_key:
            return None
        try:
            pending_store.set(pending_key, tx_hash)
        except Exception as e:
            return SettleResponse(
                success=False,
                error_reason=failed_reason,
                error_message=f"settlement_pending, but failed to persist for retry: {e}",
                transaction=tx_hash,
                network=network,
                payer=payer,
            )
        return None

    def _clear_pending() -> None:
        if pending_store is not None and pending_key:
            try:
                pending_store.delete(pending_key)
            except Exception:
                pass  # best-effort; a stale entry merely lingers until TTL expiry

    if not is_valid_tx_hash(tx_hash):
        _clear_pending()
        return invalid_broadcast_hash_response(tx_hash, failed_reason, network, payer)

    try:
        receipt = signer.wait_for_transaction_receipt(tx_hash)
    except Exception as e:
        override = _mark_pending()
        return override if override is not None else _settlement_pending_response(
            tx_hash, network, payer, e
        )

    try:
        if receipt.status != TX_STATUS_SUCCESS:
            _clear_pending()
            return SettleResponse(
                success=False,
                error_reason=failed_reason,
                transaction=tx_hash,
                network=network,
                payer=payer,
            )

        if validate_receipt is not None:
            validation_failure = validate_receipt(receipt)
            if validation_failure is not None:
                _clear_pending()
                return validation_failure

        _clear_pending()

        if on_success is not None:
            return on_success(receipt)

        return SettleResponse(
            success=True,
            transaction=tx_hash,
            network=network,
            payer=payer,
            amount=amount,
        )
    except Exception as e:
        override = _mark_pending()
        return override if override is not None else _settlement_pending_response(
            tx_hash, network, payer, e
        )


def _settlement_pending_response(
    tx_hash: str,
    network: str,
    payer: str | None,
    error: Exception,
) -> SettleResponse:
    """Build the non-terminal failure for a broadcast whose effect is unconfirmed."""
    return SettleResponse(
        success=False,
        error_reason=ERR_SETTLEMENT_PENDING,
        error_message=truncate_error_message(str(error)),
        transaction=tx_hash,
        network=network,
        payer=payer,
    )

File: typescript/packages/mechanisms/evm/src/shared/settleReceipt.ts
import { Network, SettleResponse } from "@x402/core/types";
import { PendingSettlementStore } from "@x402/core/facilitator";
import { invalidBroadcastHashResponse, isValidTxHash, truncateErrorMessage } from "../utils";
import { ErrInvalidTransactionState, ErrSettlementPending } from "../exact/facilitator/errors";
import { FacilitatorEvmSigner } from "../signer";

type SettleReceipt = Awaited<ReturnType<FacilitatorEvmSigner["waitForTransactionReceipt"]>>;

/**
 * Optional behavior for {@link waitAndReturnSettleResponse}.
 */
export interface WaitForSettleReceiptOptions {
  /**
   * Error reason for terminal failures: an invalid broadcast hash or reverted receipt.
   * Receipt-wait failures always report `ErrSettlementPending` regardless of this value.
   * Defaults to `ErrInvalidTransactionState`.
   */
  failedStatusReason?: string;
  /**
   * Optional check after a successful receipt (e.g. Transfer event). Return a SettleResponse
   * to fail settlement; return undefined to accept success.
   */
  validateReceipt?: (receipt: SettleReceipt) => SettleResponse | undefined;
  /** Settled amount attached on success when `onSuccess` is omitted. */
  amount?: string;
  /** Builds the success response from the receipt when set. */
  onSuccess?: (receipt: SettleReceipt) => SettleResponse | Promise<SettleResponse>;
}

/**
 * Waits for a transaction receipt and returns the appropriate SettleResponse.
 *
 * @param signer - Signer with waitForTransactionReceipt capability
 * @param tx - The transaction hash to wait for
 * @param network - Network the transaction was broadcast to
 * @param payer - The payer address
 * @param options - Optional receipt-wait behavior; see {@link WaitForSettleReceiptOptions}
 * @returns Promise resolving to a settlement response indicating success or failure
 */
export async function waitAndReturnSettleResponse(
  signer: Pick<FacilitatorEvmSigner, "waitForTransactionReceipt">,
  tx: `0x${string}`,
  network: Network,
  payer: string | undefined,
  options: WaitForSettleReceiptOptions = {},
): Promise<SettleResponse> {
  const {
    failedStatusReason = ErrInvalidTransactionState,
    validateReceipt,
    amount,
    onSuccess,
  } = options;

  if (!isValidTxHash(tx)) {
    return invalidBroadcastHashResponse(tx, failedStatusReason, network, payer);
  }

  let receipt;
  try {
    receipt = await signer.waitForTransactionReceipt({ hash: tx });
  } catch (error) {
    return settlementPendingResponse(tx, network, payer, error);
  }

  try {
    if (receipt.status !== "success") {
      return {
        success: false,
        errorReason: failedStatusReason,
        transaction: tx,
        network,
        payer,
      };
    }

    const validationFailure = validateReceipt?.(receipt);
    if (validationFailure) {
      return validationFailure;
    }

    if (onSuccess) {
      return await onSuccess(receipt);
    }

    return {
      success: true,
      transaction: tx,
      network,
      payer,
      ...(amount !== undefined ? { amount } : {}),
    };
  } catch (error) {
    // Processing a confirmed receipt threw, leaving the transaction onchain with an unknown
    // effect. A reverted receipt and an explicit validation failure return above and stay
    // terminal; this does not.
    return settlementPendingResponse(tx, network, payer, error);
  }
}

/**
 * Wraps a settle attempt with `PendingSettlementStore` bookkeeping: only a
 * retryable `settlement_pending` failure (carrying a broadcast transaction
 * hash) is recorded, so a subsequent settle attempt for the same payload can
 * reconcile against it instead of re-broadcasting. Every other outcome —
 * success, or any terminal failure (e.g. a reverted receipt) — clears the
 * entry instead; a terminal failure still has a `transaction` hash but is
 * not safe to reconcile against indefinitely, so it must not be cached until
 * TTL expiry.
 *
 * If persisting the pending entry itself fails, a later retry has no record
 * to reconcile against — blindly returning `settlement_pending` would let it
 * re-verify/re-broadcast and risk a double-send. That case is downgraded to
 * `nonRetryableReason`, preserving the transaction hash for manual
 * reconciliation. A failure to clear the entry (success or terminal-failure
 * path) is swallowed instead: the settle outcome is already correct and
 * must not be masked by a storage hiccup, and a stale entry merely lingers
 * until TTL expiry.
 *
 * Used by mechanisms whose failure reasons need no special-casing beyond
 * "did this attempt broadcast a transaction" (Permit2 exact/upto,
 * batch-settlement deposit, EIP-3009).
 *
 * @param store - The pending-settlement store to update
 * @param pendingKey - Deterministic key for this payload (e.g. a signature); when
 *   undefined, the store is left untouched entirely
 * @param settle - Thunk that performs the settle attempt (broadcast + receipt wait)
 * @param nonRetryableReason - Error reason to report if a `settlement_pending`
 *   outcome cannot be persisted to `store`. Defaults to `ErrInvalidTransactionState`.
 * @returns The settle result from `settle()`, unmodified unless persisting a
 *   pending entry failed
 */
export async function withPendingSettlementStore(
  store: PendingSettlementStore,
  pendingKey: string | undefined,
  settle: () => Promise<SettleResponse>,
  nonRetryableReason: string = ErrInvalidTransactionState,
): Promise<SettleResponse> {
  const result = await settle();
  if (!pendingKey) {
    return result;
  }

  const isPending =
    !result.success && result.errorReason === ErrSettlementPending && !!result.transaction;

  if (isPending) {
    try {
      await store.set(pendingKey, result.transaction as string);
    } catch (storeError) {
      return {
        ...result,
        errorReason: nonRetryableReason,
        errorMessage: `settlement_pending, but failed to persist for retry: ${
          storeError instanceof Error ? storeError.message : String(storeError)
        }`,
      };
    }
    return result;
  }

  try {
    await store.delete(pendingKey);
  } catch {
    // Best-effort cleanup; see doc comment above.
  }
  return result;
}

/**
 * Builds the non-terminal failure for a broadcast whose effect could not be established,
 * preserving the hash for the caller to reconcile against.
 *
 * @param tx - The broadcast transaction hash
 * @param network - Network the transaction was broadcast to
 * @param payer - The payer address
 * @param error - The receipt-wait or receipt-processing error
 * @returns Failed settle response with reason `settlement_pending`
 */
function settlementPendingResponse(
  tx: `0x${string}`,
  network: Network,
  payer: string | undefined,
  error: unknown,
): SettleResponse {
  return {
    success: false,
    errorReason: ErrSettlementPending,
    errorMessage: truncateErrorMessage(error instanceof Error ? error.message : String(error)),
    transaction: tx,
    network,
    payer,
  };
}
